engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java

```diff
@@ -0,0 +1,276 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.dag.gsoc;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.assets.ResourceUrn;
+import org.terasology.context.Context;
+import org.terasology.engine.SimpleUri;
+import org.terasology.engine.module.ModuleManager;
+import org.terasology.naming.Name;
+import org.terasology.rendering.assets.material.Material;
+import org.terasology.rendering.dag.Node;
+import org.terasology.rendering.dag.StateChange;
+import org.terasology.rendering.opengl.BaseFBOsManager;
+import org.terasology.rendering.opengl.FBO;
+import org.terasology.rendering.opengl.FBOConfig;
+import org.terasology.utilities.Assets;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This class implements a number of default methods for the convenience of classes
+ * wishing to implement the Node interface.
+ * <p>
+ * It provides the default functionality to identify a node, handle its status (enabled/disabled),
+ * deal with StateChange objects and Frame Buffer objects.
+ */
+public abstract class NewAbstractNode implements Node {
+    protected static final Logger logger = LoggerFactory.getLogger(NewAbstractNode.class);
+
+    protected boolean enabled = true;
+    private Set<StateChange> desiredStateChanges = Sets.newLinkedHashSet();
+    private Map<SimpleUri, BaseFBOsManager> fboUsages = Maps.newHashMap();
+    private Map<String, DependencyConnection> inputConnections = Maps.newHashMap();
+    private Map<String, DependencyConnection> outputConnections = Maps.newHashMap();
+    private final SimpleUri nodeUri;
+
+    /**
+     * Constructor to be used by inheriting classes.
+     * <p>
+     * The nodeId provided in input will become part of the nodeUri uniquely identifying the node in the RenderGraph,
+     * i.e. "engine:hazeNode".
+     *
+     * @param nodeId  a String representing the id of the node, namespace -excluded-: that's added automatically.
+     * @param context a Context object.
+     *////TODO LIST OF INS AND OUTS
+    protected NewAbstractNode(String nodeId, Context context) {
+        ModuleManager moduleManager = context.get(ModuleManager.class);
+        Name providingModule = moduleManager.getEnvironment().getModuleProviding(this.getClass());
+
+        this.nodeUri = new SimpleUri(providingModule.toString() + ":" + nodeId);
+        //TODO Check for empty list of either in or out
+    }
+
+    /**
+     * Each node must implement this method to be called when the node is fully connected.
+     * This method should call other private node methods to use dependencies.
+     */
+    public abstract void setDependencies(Context context);
```

Comment:
A bypass for using up dependencies after nodes are 1.created,2.dependencies connected